tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/Counterexample.scala

```diff
@@ -0,0 +1,47 @@
+package at.forsyte.apalache.tla.bmcmt
+
+import at.forsyte.apalache.tla.lir.TlaEx
+import at.forsyte.apalache.tla.lir.TlaModule
+import at.forsyte.apalache.io.lir.ItfCounterexampleWriter
+import at.forsyte.apalache.tla.bmcmt.trex.DecodedExecution
+
+/**
+ * Representation of a counterexample found while model checking
+ *
+ * @param rootModule
+ *   The checked TLA+ module.
+ * @param states
+ *   The states leading up to the invariant violation.
+ * @param invViolated
+ *   The invariant violation to record in the counterexample. Pass
+ *   - for invariant violations: the negated invariant,
+ *   - for deadlocks: `ValEx(TlaBool(true))`,
+ *   - for trace invariants: the applied, negated trace invariant
+ *
+ * @author
+ *   Shon Feder
+ */
+case class Counterexample(module: TlaModule, states: Counterexample.States, invViolated: TlaEx)
+
+object Counterexample {
```

Comment:
This looks to me like an adapter from some internal data structures of the model checker to `Counterexample`. So maybe it should not be a companion object of `Counterexample`?

Reply:
I don't think I follow. The `apply` method is a constructor of this class, and doesn't that usually belong in a companion object?

The other two methods are implicts that work exclusively on `Counterecample`.

Could you help explain why you think it doesn't belong with the `Counterexample` or where you think it might better belong?